libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/ui/SalesforceWebViewClientHelper.java

```diff
@@ -119,39 +121,56 @@ public static boolean hasCachedAppHome(Context ctx) {
     	return cachedAppHomeUrl != null && (new File(cachedAppHomeUrl)).exists();
     }
 
-    /**
-     * Whether the given URL is one of the expected URLs used in the bootstrapping process
-     * of the app.  Used for determining the app's "home page" URL.
-     * @param url The URL to compare against the reserved list.
-     * @return True if this URL is used in the bootstrapping process, false otherwise.
-     */
     private static boolean isReservedUrl(String url) {
-        if (url == null || url.trim().equals(""))
+        if (url == null || url.trim().equals("")) {
             return false;
+        }
         for (String reservedUrlPattern : RESERVED_URL_PATTERNS) {
-            if (url.toLowerCase(Locale.US).contains(reservedUrlPattern.toLowerCase(Locale.US)))
+            if (url.toLowerCase(Locale.US).contains(reservedUrlPattern.toLowerCase(Locale.US))) {
                 return true;
+            }
         }
         return false;
     }
 
-    /**
-     * Login redirect are of the form https://host/?ec=30x&startURL=xyz
-     * @param url
-     * @return null if this is not a login redirect and return the the value for startURL if this is a login redirect
-     */
-    private static String isLoginRedirect(String url) {
-    	final Uri uri = Uri.parse(url);
+    private static String isLoginRedirect(Context ctx, String url) {
+        final Uri uri = Uri.parse(url);
         final Map<String, String> params = UriFragmentParser.parse(uri);
     	final String ec = params.get("ec");
     	int ecInt = (ec != null ? Integer.parseInt(ec) : -1);
     	final String startURL = params.get("startURL");
-        if ((ecInt == HttpURLConnection.HTTP_MOVED_PERM
-    			|| ecInt == HttpURLConnection.HTTP_MOVED_TEMP)
-    			&& startURL != null) {
-    		return startURL;
-    	} else {
+        if (ecInt == HttpURLConnection.HTTP_MOVED_PERM
+    			|| ecInt == HttpURLConnection.HTTP_MOVED_TEMP) {
+            if (!TextUtils.isEmpty(startURL)) {
+                return startURL;
+            } else {
+                return BootConfig.getBootConfig(ctx).getStartPage();
+            }
+        } else if (isSamlLoginRedirect(ctx, url)) {
+            return BootConfig.getBootConfig(ctx).getStartPage();
+        } else {
     		return null;
     	}
     }
+
+    private static boolean isSamlLoginRedirect(Context ctx, String url) {
+        if (ctx instanceof SalesforceDroidGapActivity) {
+            final AuthConfigUtil.MyDomainAuthConfig authConfig = ((SalesforceDroidGapActivity) ctx).getAuthConfig();
+            if (authConfig != null) {
+                final List<String> ssoUrls = authConfig.getSsoUrls();
+                if (ssoUrls != null && ssoUrls.size() > 0) {
+                   for (String ssoUrl : ssoUrls) {
+                       int paramsIndex = ssoUrl.indexOf("?");
```

Comment:
The SSO URL itself could have additional crap in it. All we care about is the base URL + path.

Reply:
Couldn't we use https://developer.android.com/reference/android/net/Uri?

Reply:
I could, but that would involve more work to parse the string, append base + path, etc. This seemed simpler because we get the URL as a string and all we want to do is chop off any params if they exist.